gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,594 @@
+        parent_pipe, child_pipe = self.ctx.Pipe()
+        process = self.ctx.Process(target=target,
+            name='Worker<{0}>-{1}'.format(type(self).__name__, index),
+            args=(index, CloudpickleWrapper(self.env_fns[index]), child_pipe,
+            parent_pipe, self._obs_buffer, self.error_queue, self.episodic))
+
+        process.deamon = True
+        process.start()
+        child_pipe.close()
+
+        return process, parent_pipe
+
+    def __del__(self):
+        if hasattr(self, 'closed'):
+            if not self.closed:
+                self.close(terminate=True)
+
+
+def _worker(index, env_fn, pipe, parent_pipe, shared_memory, error_queue,
+            episodic):
+    assert shared_memory is None
+    env = env_fn()
+    _zero_observation = create_empty_array(env.observation_space, n=None,
+                                           fn=np.zeros)
+    episode_done = False
+    parent_pipe.close()
+    try:
+        while True:
+            command, data = pipe.recv()
+            if command == 'reset':
+                observation = env.reset()
+                episode_done = False
+                pipe.send(observation)
+            elif command == 'step':
+                if episodic and episode_done:
+                    observation = _zero_observation
+                    reward, done, info = 0., True, {}
+                else:
+                    observation, reward, done, info = env.step(data)
+                    if done:
+                        if episodic:
+                            observation = _zero_observation
+                            episode_done = True
+                            info.update({'AsyncVectorEnv.end_episode': True})
+                        else:
+                            observation = env.reset()
+                pipe.send((observation, reward, done, info))
+            elif command == 'seed':
+                env.seed(data)
+                pipe.send(None)
+            elif command == 'close':
+                pipe.send(None)
+                break
+            elif command == '_call':
+                name, args, kwargs = data
+                if name in ['reset', 'step', 'seed', 'close']:
+                    raise ValueError('Trying to call function `{0}` with '
+                        '`_call`. Use `{0}` directly instead.'.format(name))
+                function = getattr(env, name)
+                if callable(function):
+                    pipe.send(function(*args, **kwargs))
+                else:
+                    pipe.send(function)
+            elif command == '_setattr':
+                name, value = data
+                setattr(env, name, value)
+                pipe.send(None)
+            elif command == '_check_observation_space':
+                pipe.send(data == env.observation_space)
+            elif command == '_restart':
+                observation = env.reset()
+                episode_done = True
```

Comment:
why is `episode_done = True` here? It seems to me that after restart (and calling env.reset()) it should be False

Reply:
This makes sense, it should probably only set `episode_done = True` if we are calling `step`.